### PR TITLE
Feature/issue 1490 loop order

### PR DESCRIPTION
**Pull request by peterwicksstringfield**
## Summary

Partially addresses #1490.

Eliminates instances of row-major iteration across a column-major Eigen matrix in six places:
stan/math/prim/mat/prob/lkj_cov_lpdf.hpp
stan/math/prim/mat/prob/matrix_normal_prec_rng.hpp
stan/math/fwd/mat/fun/to_fvar.hpp
stan/math/fwd/mat/fun/mdivide_left_ldlt.hpp
stan/math/fwd/mat/fun/inverse.hpp
stan/math/rev/mat/functor/kinsol_data.hpp

I can trivially break this into multiple PRs if that would make it easier.

This pattern occurs frequently when handling matrices of fvars (especially in the mdivide_foo family):
```
for (int j = 0; j < b.cols(); j++) {
  for (int i = 0; i < b.rows(); i++) {
       b_val(i, j) = b(i, j).val_;
       b_der(i, j) = b(i, j).d_;
```

There were three places where it was row-col instead of col-row. I fixed that.

It could be:
```
b_val = b.val();
b_der = b.d();
```

Which would be two walks across b. Well, technically val and d both return CWiseUnaryOps, and the walks happen when they are turned into dense matrices.

kinsol_data is tested indirectly, via its uses in the algebra solvers.

In Stan, lkj_cov_lpdf is depreciated in favor of lkj_corr_lpdf. Also lkj_cov_lpdf had no tests. The file that was supposed to contain its tests just had a copy of the lkj_corr_lpdf tests. I've added a few tests for error conditions in lkj_cov_lpdf, but its still not proper coverage.

When reviewing the lkj tests, note the following: if y is a covariance matrix, and D is a diagonal matrix with Dii = 1/sqrt[yii], then DyD is the associated correlation matrix. Now if yii = 1, mu_i = 0, sd_i = 1, then:
```
lkj_cov_lpdf[y, mu, sigma, eta]
    ~= lkj_corr_lpdf[DyD, eta] + sum_i lognormal_lpdf[sqrt[yii], mu_i, sigma_i]
    = lkj_corr_lpdf[y, eta] + sum_i lognormal_lpdf[1, 0, 1]
    = lkj_corr_lpdf[y, eta] + size_of_y * lognormal_lpdf[1, 0, 1]
```

## Tests

Added tests:
test/unit/math/prim/mat/prob/lkj_cov_test.cpp

Old tests:
test/unit/math/mix/mat/fun/inverse_test.cpp
test/unit/math/prim/mat/fun/inverse_test.cpp
test/unit/math/prim/mat/prob/matrix_normal_prec_rng_test.cpp
test/unit/math/mix/mat/fun/to_fvar_test.cpp
test/unit/math/fwd/mat/fun/to_fvar_test.cpp
test/unit/math/mix/mat/fun/mdivide_left_ldlt_test.cpp
test/unit/math/prim/mat/fun/mdivide_left_ldlt_test.cpp
test/unit/math/rev/mat/functor/algebra_solver_fp_test.cpp
test/unit/math/rev/mat/functor/algebra_solver_test.cpp

## Side Effects

None.

## Checklist

- [X] Math issue #1490

- [X] Copyright holder: Peter Wicks Stringfield

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [X] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [X] the code is written in idiomatic C++ and changes are documented in the doxygen

- [X] the new changes are tested

### CODE REVIEW

File: stan/math/fwd/mat/fun/inverse.hpp [hunk 1]
```diff
@@ -15,12 +15,13 @@ namespace math {
 template <typename T, int R, int C>
 inline Eigen::Matrix<fvar<T>, R, C> inverse(
     const Eigen::Matrix<fvar<T>, R, C>& m) {
+  check_nonempty("inverse", "m", m);
```
Comment (rok-cesnovar) [inline stan/math/fwd/mat/fun/inverse.hpp]: @mcol you are working on these a lot. Just double checking with you if this is what we want or would returning an empty matrix be better. I do think we should check that the matrix is square before the nonempty check.

Comment (mcol) [reply stan/math/fwd/mat/fun/inverse.hpp]: For inverse we always throw if it's empty, so the order of the checks doesn't quite matter (see #1464 for example).

Comment (rok-cesnovar) [reply stan/math/fwd/mat/fun/inverse.hpp]: Sweet. All good then. Thanks @mcol

Comment (peterwicksstringfield) [reply stan/math/fwd/mat/fun/inverse.hpp]: Looks like you've figured it out, but that function threw on non empty anyway. Inverse<fwd<T> > calls inverse<T>, which eventually terminates at either inverse<prim> or inverse<var>. Prim and var threw on non-empty, so fwd did too. This change just makes it explicit.

**Pull request merged**